### PR TITLE
[exec] Add exec  plugin

### PR DESCRIPTION
**Pull request by elfiesmelfie**
Add template for the exec plugin

### CODE REVIEW

**Comment by elfiesmelfie** [comment]
This is used by tripleo_collectd_ansible_role for the sensubility plugin

https://github.com/infrawatch/tripleo-collectd-ansible-role/tree/efoley-tripleo_testing_updates

File: templates/exec.conf.j2 [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+{% else %}
+LoadPlugin exec
+{% endif %}
+
+<Plugin "exec">
+{% for exec in collectd_plugin_exec_exec %}
+  Exec "{{ exec.user }}{% if exec.group is defined %}:{{ exec.group}}{% endif %}" "{{ exec.exec }}{% for arg in exec.args | default([]) %} arg{% endfor %}"
+{% endfor %}
+{% for exec in collectd_plugin_exec_notification %}
+  Exec "{{ exec.user }}{% if exec.group is defined %}:{{ group}}{% endif %}" "{{ exec.notification_exec }}{% for arg in exec.args %} arg{% endfor %}"
```
Comment (leifmadsen) [inline templates/exec.conf.j2]:
```suggestion
  NotificationExec "{{ exec.user }}{% if exec.group is defined %}:{{ group}}{% endif %}" "{{ exec.notification_exec }}{% for arg in exec.args %} arg{% endfor %}"
```

File: templates/exec.conf.j2 [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+<LoadPlugin exec>
+    Globals {{ collectd_plugin_exec_globals }}
+</LoadPlugin>
+{% else %}
+LoadPlugin exec
+{% endif %}
+
+<Plugin "exec">
+{% for exec in collectd_plugin_exec_exec %}
+  Exec "{{ exec.user }}{% if exec.group is defined %}:{{ exec.group}}{% endif %}" "{{ exec.exec }}{% for arg in exec.args | default([]) %} arg{% endfor %}"
```
Comment (leifmadsen) [inline templates/exec.conf.j2]: Should the `arg` be `{{ arg }}` ?

File: defaults/main/exec.yml [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+# - user: collectd
+#   group: collectd
+#   exec: "my-prog"
+#   args: "arg0" arg1"
+
+collectd_plugin_exec_notification: []
+# collectd_plugin_exec_notification:
+# - user: collectd
+#   group: collectd
+#   notification_exec: "my-notifier"
```
Comment (leifmadsen) [inline defaults/main/exec.yml:13]: Should you also provide the `args` parameter example here?

File: defaults/main/exec.yml [hunk 4]
```diff
@@ -0,0 +1,16 @@
+---
+collectd_plugin_exec_exec: []
+# collectd_plugin_exec_exec:
+# - user: collectd
+#   group: collectd
+#   exec: "my-prog"
+#   args: "arg0" arg1"
```
Comment (leifmadsen) [inline defaults/main/exec.yml]: Comments are not completed properly. Should be `"arg0" "arg1"` or `"arg0 arg1"` (I'm not sure which).

File: templates/intel_pmu.conf.j2 [hunk 5]
```diff
@@ -10,8 +10,9 @@ LoadPlugin intel_pmu
 {% if collectd_plugin_intelpmu_cores is defined %}
   Cores{% for item in collectd_plugin_intelpmu_cores %} "{{ item }}"{% endfor %}
 {% endif %}
-
+{% if collectd_plugin_intelpmu_reporthardwarecacheevents is defined %}
   ReportHardwareCacheEvents {{ collectd_plugin_intelpmu_reporthardwarecacheevents | string | lower }}
+{% endif %}
   ReportKernelPMUEvents {{ collectd_plugin_intelpmu_reportkernelpmuevents | string | lower }}
```
Comment (leifmadsen) [inline templates/intel_pmu.conf.j2:17]: I think you want to wrap this in the same `is defined` no?  https://github.com/voxpupuli/puppet-collectd/blob/master/templates/plugin/intel_pmu.conf.erb#L5-L7

File: templates/intel_pmu.conf.j2 [hunk 6]
```diff
@@ -10,8 +10,9 @@ LoadPlugin intel_pmu
...
   ReportSoftwareEvents {{ collectd_plugin_intelpmu_reportsoftwareevents | string | lower }}
```
Comment (leifmadsen) [inline templates/intel_pmu.conf.j2:20]: Same as above, `is defined` wrapper?  https://github.com/voxpupuli/puppet-collectd/blob/master/templates/plugin/intel_pmu.conf.erb#L8-L10

File: templates/intel_pmu.conf.j2 [hunk 7]
```diff
@@ -10,8 +10,9 @@ LoadPlugin intel_pmu
...
 {% if collectd_plugin_intelpmu_eventlist is defined %}
```
Comment (leifmadsen) [inline templates/intel_pmu.conf.j2:22]: Couple comments about the below section (I know not part of this review...)

The event_list seems like it should be a list of all the events in a single argument surrounded by double quotes?

* https://github.com/voxpupuli/puppet-collectd/blob/master/templates/plugin/intel_pmu.conf.erb#L11-L13

Also the HardwareEvents looks like it should be a single argument in double quotes, with all list items joined by a comma

* https://github.com/voxpupuli/puppet-collectd/blob/master/templates/plugin/intel_pmu.conf.erb#L14-L16

File: templates/exec.conf.j2 (hunk 2 above)
Comment (elfiesmelfie) [reply templates/exec.conf.j2]: Yes it should

File: defaults/main/exec.yml (hunk 3 above)
Comment (elfiesmelfie) [reply defaults/main/exec.yml:13]: The args example needs to be something different.

File: defaults/main/exec.yml (hunk 4 above)
Comment (elfiesmelfie) [reply defaults/main/exec.yml]: The args example needs to be something different.

File: templates/intel_pmu.conf.j2 (hunk 7 above)
Comment (elfiesmelfie) [reply templates/intel_pmu.conf.j2:22]: I'll have to take a close look at that, I've added issue https://github.com/infrawatch/collectd-config-ansible-role/issues/29 to track this and will probably pull this template into a separate branch/PR.

**Review by csibbitt — APPROVED** [review]

**Pull request merged**